rllib/env/multi_agent_env_runner.py

```diff
@@ -0,0 +1,887 @@
+            )
+
+            # TODO (simon): The `gym.Wrapper` for `gym.vector.VectorEnv` should
+            #  actually hold the spaces for a single env, but for boxes the
+            #  shape is (1, 1) which brings a problem with the action dists.
+            #  shape=(1,) is expected.
+            module_spec.action_space = self.env.action_space
+            module_spec.observation_space = self.env.observation_space
+            # Set action and observation spaces for all module specs.
+            for agent_id, agent_module_spec in module_spec.module_specs.items():
+                # Note, `MultiAgentEnv` has a preferred format of spaces, i.e.
+                # a mapping from agent ids to spaces.
+                # If the action space is a mapping from agent ids to spaces.
+                if self._action_space_in_preferred_format:
+                    agent_module_spec.action_space = self.env.action_space[agent_id]
+                # Otherwise, use the same space for each agent.
+                else:
+                    agent_module_spec.action_space = self.env.action_space
+                # Similar to the action spaces, observation spaces could be described
+                # by a mapping.
+                if self._obs_space_in_preferred_format:
+                    agent_module_spec.observation_space = self.env.observation_space[
+                        agent_id
+                    ]
+                # Otherwise, use the same space for all agents.
+                else:
+                    agent_module_spec.observation_space = self.env.observation_space
+
+            # Build the module from its spec.
+            self.marl_module: MultiAgentRLModule = module_spec.build()
+            # Get the policy mapping function b/c we need it to map observations and
+            # states to single-agent `RLModule`s.
+            # TODO (sven, simon): We have to rebuild the `AlgorithmConfig` to work on
+            # `RLModule`s and not `Policy`s. Like here `policies`->`modules`
+            self._agent_id_to_module_id = {
+                agent_id: self.config.policy_mapping_fn()(agent_id, None, self)
+                for agent_id in self.agent_ids
+            }
+            # Remove as soon as multi-agent connectors are ready.
+            self._module_id_to_agent_id = {
+                module_id: agent_id
+                for agent_id, module_id in self._agent_id_to_module_id.items()
+            }
+        # This error could be thrown, when only random actions are used.
+        except NotImplementedError:
+            self.marl_module = None
+
+        # This should be the default.
+        self._needs_initial_reset: bool = True
+        self._episode: "MultiAgentEpisode" = None
+
+        self._done_episodes_for_metrics: List["MultiAgentEpisode"] = []
+        self._ongoing_episode_for_metrics: Dict[
+            List["MultiAgentEpisode"]
+        ] = defaultdict(list)
+        self._ts_since_last_metrics: int = 0
```

Comment:
I think this is not even needed. Was the same for SingleAgentEpisode.
Let's remove this property (_ts_since_last_metrics).

Reply:
done